pkg/bpf/bpf_linux.go

```diff
@@ -460,82 +354,55 @@ func objCheck(fd int, path string, mapType MapType, keySize, valueSize, maxEntri
 	return false
 }
 
-func OpenOrCreateMap(path string, mapType MapType, keySize, valueSize, maxEntries, flags uint32, innerID uint32, pin bool) (int, error) {
-	var fd int
-	var err error
-
-	redo := false
-
-recreate:
-	create := true
-	if pin {
-		if _, err := os.Stat(path); os.IsNotExist(err) || redo {
-			mapDir := filepath.Dir(path)
-			if _, err = os.Stat(mapDir); os.IsNotExist(err) {
-				if err = os.MkdirAll(mapDir, 0755); err != nil {
-					return 0, &os.PathError{
-						Op:   "Unable create map base directory",
-						Path: path,
-						Err:  err,
-					}
-				}
-			}
-		} else {
-			create = false
+// OpenOrCreateMap attempts to load the pinned map at "pinDir/<spec.Name>" if
+// the spec is marked as Pinned. Any parent directories of pinDir are
+// automatically created. Any pinned maps incompatible with the given spec are
+// removed and recreated.
+//
+// If spec.Pinned is 0, a new Map is always created.
+func OpenOrCreateMap(spec *ebpf.MapSpec, pinDir string) (*ebpf.Map, error) {
+	var opts ebpf.MapOptions
+	if spec.Pinning != 0 {
+		if pinDir == "" {
+			return nil, errors.New("cannot pin map to empty pinDir")
+		}
+		if spec.Name == "" {
+			return nil, errors.New("cannot load unnamed map from pin")
+		}
+
+		if err := os.MkdirAll(pinDir, 0755); err != nil {
+			return nil, fmt.Errorf("creating map base pinning directory: %w", err)
 		}
-	}
 
-	if create {
-		fd, err = CreateMap(
-			mapType,
-			keySize,
-			valueSize,
-			maxEntries,
-			flags,
-			innerID,
-			path,
-		)
-
-		defer func() {
-			if err != nil {
-				// In case of error, we need to close
-				// this fd since it was open by CreateMap
-				ObjClose(fd)
-			}
-		}()
+		opts.PinPath = pinDir
+	}
 
+	m, err := createMap(spec, &opts)
+	if errors.Is(err, ebpf.ErrMapIncompatible) {
+		// Found incompatible map. Open the pin again to find out why.
+		m, err := ebpf.LoadPinnedMap(path.Join(pinDir, spec.Name), nil)
 		if err != nil {
-			return 0, err
+			return nil, fmt.Errorf("open pin of incompatible map: %w", err)
 		}
-
-		if pin {
-			err = ObjPin(fd, path)
-			if err != nil {
-				return 0, err
-			}
+		defer m.Close()
+
+		log.WithField(logfields.Path, path.Join(pinDir, spec.Name)).
+			WithFields(logrus.Fields{
+				"old": fmt.Sprintf("Type:%s KeySize:%d ValueSize:%d MaxEntries:%d, Flags:%d",
```

Comment:
This may be personal preference, but I find the kv approach harder to visually diff: https://go.dev/play/p/xlpzqnxCIig.

```
time="2009-11-10T23:00:00Z" level=info msg="Unpinning map with incompatible properties" new="Type:Array KeySize:4 ValueSize:4 MaxEntries:1024 Flags:0" old="Type:Hash KeySize:4 ValueSize:4 MaxEntries:1024 Flags:0"
```
vs
```
time="2009-11-10T23:00:00Z" level=info msg="Unpinning map with incompatible properties" new_flags=0 new_keysize=4 new_maxentries=4 new_type=Array new_valuesize=4 old_flags=0 old_keysize=4 old_maxentries=4 old_type=Hash old_valuesize=4
```

- Lots of repetition (old_ and new_ prefixes)
- Longer
- Fields get sorted alphabetically
- The convention is to lowercase keys, which sort of breaks the correspondence to the keys in the MapSpec

Reply:
I see what you mean, the main advantage of the standard way is that it might hook better into existing log aggregation services.